### PR TITLE
Close threads at server shutdown

### PR DESCRIPTION
**Pull request by petersv5**
To exit normally all non-daemon threads of the java process are supposed to have stopped. There are several cases of EventLoopGroup-created thread pools that are not shut down as a part of the server shutdown.

These threads are spawned on the first Bedrock player login. The identified cases are:
- TcpClientSocket created by GeyserSession
- The listener threads in GeyserServerInitializer
- The player threads in GeyserServerInitializer
- The EXECUTOR_SERVICE in the SkinProvider

This patch depends on the PR
https://github.com/GeyserMC/MCProtocolLib/pull/744 submitted to MCProtocolLib.

This is a very quickly thrown together proof of concept to verify that the execution thread pools left running were what blocked a clean exit with Geyser active.

### CODE REVIEW

File: core/src/main/java/org/geysermc/geyser/network/netty/GeyserServer.java [hunk 1]
```diff
@@ -85,6 +87,7 @@ public final class GeyserServer {
     private final GeyserImpl geyser;
     private final EventLoopGroup group;
     private final ServerBootstrap bootstrap;
+    private EventLoopGroup player_group;
```
Comment (Camotoy) [inline core/src/main/java/org/geysermc/geyser/network/netty/GeyserServer.java:90]: playerGroup

**Comment by Camotoy** [comment]
Can we make sure this works with `/geyser reload`?

**Comment by petersv5** [comment]
> Can we make sure this works with `/geyser reload`?

Will that call GeyserServer.shutdown()? If so things will break for the cases where the EventLoopGroup is held in a static class variable and some other approach will be needed.

**Comment by Konicai** [comment]
Spigot's reload should also be tested (despite it being terrible) to make sure things don't get any worse than they already are

**Comment by petersv5** [comment]
Like I wrote in the MCProtocolLib PR, it is probably best for someone that knows Geyser to reimplement these changes properly. This patch was literally created with grep, jstack and jdb.

**Comment by petersv5** [comment]
I can confirm /geyser reload is completely broken, as expected.
I'll make a version that actually unsets the variables tracking the event loop groups which should allow them to be recreated.

However, I wonder if the better solution would be to make the event loop groups instance variables instead of class members. I think we only create one instance of any of the classes holding event loop groups so there is no real point in making them static class variables. We never reuse them anyway. I need to go over all of them, but I think we never create many of them. At least not on a regular server. Are there cases with multiple instances in one process and where it is desireable to share worker pools?

**Comment by petersv5** [comment]
I pushed an update that allows /geyser reload to work. 
The patch now nullifies EventLoopGroup holders when shut down.
EventLoopGroup users that are class variables can automatically restart the EventLoopGroup on first use. This affected TcpClientSession (see the updated patch to MCProtocolLib) and SkinProvider. The remaining EventLoopGroups were instance variables of objects that were recreated after a /geyser reload.
I also changed the name of player_group to playerGroup as requested.

I am not able to test spigot reload as I do not have such a system set up anymore.

File: core/src/main/java/org/geysermc/geyser/network/netty/GeyserServer.java (hunk 1 above)
Comment (petersv5) [reply core/src/main/java/org/geysermc/geyser/network/netty/GeyserServer.java]: PR updated

**Comment by petersv5** [comment]
Rebased to latest master.

**Comment by petersv5** [comment]
Rebased to latest master

**Comment by petersv5** [comment]
Switched to using the alterantive PR for ProtocolLib (https://github.com/GeyserMC/MCProtocolLib/pull/748) after comments from Camotoy. This one no longer has an explicit shutdown of the TcpClientSession EventLoopGroup. That is now reaped on process exit.

The remaining EventLoopGroup threads are still shutdown on Geyser shutdown. 

I have tested /geyser reload and that seems to work.

**Comment by petersv5** [comment]
Rebased to latest master.

**Comment by petersv5** [comment]
Rebased

**Comment by petersv5** [comment]
Rebased and tested

**Comment by Konicai** [comment]
thank you

File: core/src/main/java/org/geysermc/geyser/skin/SkinProvider.java [hunk 2]
```diff
@@ -133,6 +133,19 @@ public class SkinProvider {
         WEARING_CUSTOM_SKULL_SLIM = new SkinGeometry("{\"geometry\" :{\"default\" :\"geometry.humanoid.wearingCustomSkullSlim\"}}", wearingCustomSkullSlim, false);
     }
 
+    private static ExecutorService getExecutorService() {
+        if (EXECUTOR_SERVICE == null) {
+            EXECUTOR_SERVICE = Executors.newFixedThreadPool(ALLOW_THIRD_PARTY_CAPES ? 21 : 14);
+        }
+        return EXECUTOR_SERVICE;
+    }
+    public static void shutdown() {
```
Comment (Konicai) [inline core/src/main/java/org/geysermc/geyser/skin/SkinProvider.java:142]: space here please

Comment (petersv5) [reply core/src/main/java/org/geysermc/geyser/skin/SkinProvider.java:143]: Updated and pushed

**Review by onebeastchris — APPROVED** [review]
Tested on all platforms (compiled locally with the mcpl pr), works perfectly! Thanks for this fix & for keeping the PR updated :)

**Comment by Konicai** [comment]
Until 1.20.2 when we can target a newer MCPL

**Review by Konicai — APPROVED** [review]

**Review by rtm516 — APPROVED** [review]

**Comment by Konicai** [comment]
thanks again!

**Pull request merged**